\documentclass[]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage[]{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\PassOptionsToPackage{hyphens}{url} % url is loaded by hyperref
\usepackage[unicode=true]{hyperref}
\hypersetup{
            pdftitle={Scissor Tutorial},
            pdfauthor={Duanchen Sun},
            pdfborder={0 0 0},
            breaklinks=true}
\urlstyle{same}  % don't use monospace font for urls
\usepackage[margin=1in]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
\IfFileExists{parskip.sty}{%
\usepackage{parskip}
}{% else
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

% set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother


\title{Scissor Tutorial}
\author{Duanchen Sun}
\date{2020-05-27}

\begin{document}
\maketitle

{
\setcounter{tocdepth}{2}
\tableofcontents
}
\section{Introduction}\label{introduction}

This Scissor R package contains the proposed Scissor algorithm (function
\texttt{Scissor}), which is a novel single-cell data analysis approach
that utilizes disease phenotypes that have been collected for bulk
assays to identify the cell subpopulations that most highly correlate
with a given phenotype. Scissor provides a novel framework to identify
the biologically and clinically relevant cell subpopulations from
single-cell assays by leveraging the wealth of phenotypes and bulk-omics
datasets. In this tutorial, we used several examples to help users
executing Scissor in real applications.

We first load in the required package:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(Scissor)}
\end{Highlighting}
\end{Shaded}

\section{Scissor implementation
example}\label{scissor-implementation-example}

\subsection{Input data preparation}\label{input-data-preparation}

There are two necessary input data sources of Scissor: a single-cell
expression matrix and a bulk profiling data with sample phenotype, which
can be a binary group indicator vector or clinical survival data. In
this tutorial, the application of lung cancer cell in our manuscript was
used as an example. First, let's load in the processed Lung
Adenocarcinoma (LUAD) TPM quantification data and the corresponding
survival information downloaded from The Cancer Genome Atlas (TCGA).

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{load}\NormalTok{(}\KeywordTok{url}\NormalTok{(}\StringTok{'https://xialab.s3-us-west-2.amazonaws.com/Duanchen/Scissor_data/bulk_dataset.RData'}\NormalTok{))}
\KeywordTok{load}\NormalTok{(}\KeywordTok{url}\NormalTok{(}\StringTok{'https://xialab.s3-us-west-2.amazonaws.com/Duanchen/Scissor_data/bulk_survival.RData'}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

In this bulk expression matrix, each row is a gene and each column is a
sample. The dimensions of LUAD bulk dataset are:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{dim}\NormalTok{(bulk_dataset)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 56716   471
\end{verbatim}

which indicate there are 56716 genes and 471 samples in total. Besides,
all of these samples have clinical outcomes:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{head}\NormalTok{(bulk_survival)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##   TCGA_patient_barcode OS_time Status
## 1         TCGA-05-4249    1158      0
## 2         TCGA-05-4250     121      1
## 3         TCGA-05-4382     607      0
## 4         TCGA-05-4384     426      0
## 5         TCGA-05-4389    1369      0
## 6         TCGA-05-4390    1126      0
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{all}\NormalTok{(}\KeywordTok{colnames}\NormalTok{(bulk_dataset) }\OperatorTok{==}\StringTok{ }\NormalTok{bulk_survival}\OperatorTok{$}\NormalTok{TCGA_patient_barcode)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] TRUE
\end{verbatim}

Note that Scissor requires a three-column format of clinical data, if
the users want to identify a cell subpopulation that are correlated with
good or bad survival. The first column contains the sample ID
(observations), which should in the same order with the columns in bulk
expression matrix. The second column is the survival time and the third
column stands for the event of interest (e.g.recurrence of cancer or
death), which is indicated by the variable ``Status'' with 0 means no
event (censored) and 1 means event.

Then, we load in the LUAD single-cell RNA-seq raw counts data:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{load}\NormalTok{(}\KeywordTok{url}\NormalTok{(}\StringTok{'https://xialab.s3-us-west-2.amazonaws.com/Duanchen/Scissor_data/sc_dataset.RData'}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

Same with the bulk expression matrix, the rows and columns are genes and
cells, respectively. The dimensions of LUAD single-cell data are:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{dim}\NormalTok{(sc_dataset)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 33694  4102
\end{verbatim}

The users do not need to keep the common genes between bulk and
single-cell expression datasets, which can be automatically achieved by
the preprocessing steps of Scissor.

\subsection{Identify the most relevent cell subpopulations by
Scissor}\label{identify-the-most-relevent-cell-subpopulations-by-scissor}

Given the above inputs, now we can use Scissor to select the most
relevent cell subpopulations with the guidence of phenotype (survival
outcomes), which is fitted by a Cox regression model
(\texttt{family\ =\ \textquotesingle{}cox\textquotesingle{}}). We set
the ratio between the L1 norm and Laplacian regularization to 0.05
(\texttt{alpha\ =\ 0.05}):

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{infos <-}\StringTok{ }\KeywordTok{Scissor}\NormalTok{(}\DataTypeTok{alpha =} \FloatTok{0.05}\NormalTok{, }\DataTypeTok{family =} \StringTok{'cox'}\NormalTok{, }\DataTypeTok{Given_Inputs =} \OtherTok{NULL}\NormalTok{, }
                 \DataTypeTok{bulk_dataset =}\NormalTok{ bulk_dataset, }\DataTypeTok{sc_dataset =}\NormalTok{ sc_dataset, }\DataTypeTok{cor_type =} \StringTok{'pearson'}\NormalTok{,}
                 \DataTypeTok{survival_info =}\NormalTok{ bulk_survival, }\DataTypeTok{save_file =} \StringTok{'test.RData'}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "Scissor output information: Positive cells: 201. Negative cells: 4."
\end{verbatim}

The output message of Scissor indicates that there are 201 Scissor+
cells and 4 Scissor- cells. The information of selected cells is saved
in variable \texttt{infos}:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{names}\NormalTok{(infos)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "Coefs"         "Cell_positive" "Cell_negative" "Seurat_data"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{length}\NormalTok{(infos}\OperatorTok{$}\NormalTok{Cell_positive)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 201
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{infos}\OperatorTok{$}\NormalTok{Cell_positive[}\DecValTok{1}\OperatorTok{:}\DecValTok{4}\NormalTok{]}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "AAAGTAGAGGAGCGAG_19" "AACCATGCATCTCCCA_19" "AACCGCGAGCTGCGAA_20" "AACTCAGTCCGCGGTA_19"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{length}\NormalTok{(infos}\OperatorTok{$}\NormalTok{Cell_negative)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 4
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{infos}\OperatorTok{$}\NormalTok{Cell_negative}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "ACGCCAGTCCTCCTAG_20" "ACGGGCTAGTGGCACA_20" "CCGGTAGGTACCCAAT_15" "GACGCGTAGTGGTCCC_20"
\end{verbatim}

Besides, the output of Scissor also contains a preprocessed single-cell
Seurat object \texttt{Seurat\_data}, which can directly accept the
functions in \texttt{Seurat} package for analysis and visualization. The
internal function in Scissor will automatically save the selected cells
into a metadata in \texttt{Seurat\_data}. Therefore, users can visualize
the selected cells as:

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{DimPlot}\NormalTok{(}\DataTypeTok{object =}\NormalTok{ infos}\OperatorTok{$}\NormalTok{Seurat_data, }\DataTypeTok{reduction =} \StringTok{'umap'}\NormalTok{, }\DataTypeTok{group.by =} \StringTok{'scissor'}\NormalTok{, }
        \DataTypeTok{cols =} \KeywordTok{c}\NormalTok{(}\StringTok{'grey'}\NormalTok{,}\StringTok{'red'}\NormalTok{,}\StringTok{'blue'}\NormalTok{), }\DataTypeTok{pt.size =} \DecValTok{1}\NormalTok{, }\DataTypeTok{order =} \KeywordTok{c}\NormalTok{(}\DecValTok{2}\NormalTok{,}\DecValTok{1}\NormalTok{))}
\end{Highlighting}
\end{Shaded}

\begin{center}\includegraphics{Scissor_Tutorial_files/figure-latex/unnamed-chunk-9-1} \end{center}

in which \texttt{1} stand for the Scissor+ cells and \texttt{2} stand
for the Scissor- cells.

\subsection{Parameter tuning in
Scissor}\label{parameter-tuning-in-scissor}

Usually, users cannot obtain the most informative selected cell at the
first trial. In applications, Scissor will automatically save the
preprocessed inputs for regression into a RData
(\texttt{save\_file\ =\ \textquotesingle{}test.RData}), which is
convenient for users to directly execute the parameter tuning step.
Suppose users use the above file name \texttt{test.RData}, Scissor can
reselect the cells with a new alpha (\texttt{alpha\ =\ 0.01}) as:

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{infos <-}\StringTok{ }\KeywordTok{Scissor}\NormalTok{(}\DataTypeTok{alpha =} \FloatTok{0.01}\NormalTok{, }\DataTypeTok{family =} \StringTok{'cox'}\NormalTok{, }\DataTypeTok{Given_Inputs =} \StringTok{'test.RData'}\NormalTok{,}
\NormalTok{                 bulk_dataset, sc_dataset, bulk_survival)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "Scissor output information: Positive cells: 573. Negative cells: 20."
\end{verbatim}

A smaller alpha leads to a less sparse selection result. At this time,
the output message of Scissor indicates that there are 573 Scissor+
cells and 20 Scissor- cells. Usually, we recommand that the total
selected cells of Scissor should not exceed the minimum number of 10\%
of total cells in single-cell RNA-seq data and 1000. In practice, users
can change the alpha with \texttt{Given\_Inputs} equals to the saved
file name to meet their different goals.

\section{Reference}\label{reference}

Duanchen Sun and Zheng Xia (2020): Phenotype-guided subpopulation
identification from single-cell sequencing data

\end{document}
